ui/src/text_annotator.tsx

```diff
@@ -173,26 +176,63 @@ export
           currentTag = tag
           return arr
         }, [] as TextAnnotatorItem[]) as unknown as Rec[]
-        if (model.trigger) wave.push()
+        if (trigger) wave.push()
       },
-      activateTag = React.useCallback((tagName: S) => () => setActiveTag(tagName), [setActiveTag]),
+      activateTag = (tagName: S) => () => setActiveTag(tagName),
       removeAnnotation = (idx: U) => (ev: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
         ev.stopPropagation() // Stop event bubbling so that annotate is not called.
         const tagToRemove = tokens[idx].tag
         for (let i = idx; tokens[i]?.tag === tagToRemove; i++) tokens[i].tag = undefined
         setTokens([...tokens])
         submitWaveArgs()
       },
-      updateStartIdx = (startIdx: U) => () => setStartIdx(startIdx),
-      annotate = (end: U) => () => {
-        const start = startIdx
-        if (start === null) return
+      timeoutRef = React.useRef<NodeJS.Timeout | null>(null),
+      mouseDownTimeRef = React.useRef<number>(0),
+      annotate = (target?: EventTarget & HTMLSpanElement) => {
+        if (timeoutRef.current) {
+          clearTimeout(timeoutRef.current)
+          timeoutRef.current = null
+        }
+        const
+          selectedText = window.getSelection(),
+          startElData = selectedText?.anchorNode?.parentElement?.dataset,
+          endElData = selectedText?.focusNode?.parentElement?.dataset
+
+        if (!startElData || !endElData) return
 
-        const max = Math.max(start, end)
-        for (let i = Math.min(start, end); i <= max; i++) tokens[i].tag = activeTag
+        const
+          selectedTextStr = selectedText.toString(),
+          startTokenIdx = parseInt(smart_selection ? startElData!.key! : startElData!.start!),
+          endTokenIdx = target ? (startTokenIdx + selectedTextStr.length) : parseInt(smart_selection ? endElData!.key! : endElData!.end!),
+          max = Math.max(startTokenIdx, endTokenIdx),
+          min = Math.min(startTokenIdx, endTokenIdx)
 
+        if (target && selectedTextStr.length) {
+          // smart selection on double click
+          const targetDataIdx = parseInt(target.dataset.key!)
+          let strLength = 0 // number of characters right to the selection target; if not 0, we loop to the left direction
+          for (let i = 0; i <= selectedTextStr.length; i++) {
+            const tokenIdx = targetDataIdx + (!strLength ? i : (strLength - i))
```

Comment:
Boolean negation makes conditionals hard to read, a lot of times simple if/else branch switch is enough.
```suggestion
            const tokenIdx = targetDataIdx + (strLength ? (strLength - i) : i)
```